tests/components/demo/test_update.py

```diff
@@ -0,0 +1,153 @@
+    )
+    assert state.attributes[ATTR_DEVICE_CLASS] == UpdateDeviceClass.FIRMWARE
+
+    state = hass.states.get("update.demo_update_with_progress")
+    assert state
+    assert state.state == STATE_ON
+    assert state.attributes[ATTR_TITLE] == "Philips Lamps Firmware"
+    assert state.attributes[ATTR_CURRENT_VERSION] == "1.93.3"
+    assert state.attributes[ATTR_LATEST_VERSION] == "1.94.2"
+    assert state.attributes[ATTR_RELEASE_SUMMARY] == "Added support for effects"
+    assert (
+        state.attributes[ATTR_RELEASE_URL] == "https://www.example.com/release/1.93.3"
+    )
+    assert state.attributes[ATTR_DEVICE_CLASS] == UpdateDeviceClass.FIRMWARE
+
+
+async def test_update_with_progress(hass: HomeAssistant) -> None:
+    """Test update with progress."""
+    state = hass.states.get("update.demo_update_with_progress")
+    assert state
+    assert state.state == STATE_ON
+    assert state.attributes[ATTR_IN_PROGRESS] is False
+
+    events = []
+    hass.helpers.event.async_track_state_change_event(
+        "update.demo_update_with_progress", callback(lambda event: events.append(event))
+    )
+
+    with patch("homeassistant.components.demo.update.asyncio.sleep") as fake_sleep:
+        await hass.services.async_call(
+            DOMAIN,
+            SERVICE_INSTALL,
+            {ATTR_ENTITY_ID: "update.demo_update_with_progress"},
+            blocking=True,
+        )
+
+    assert fake_sleep.call_count == 10
+    assert len(events) == 10
+    assert events[0].data["new_state"].state == STATE_ON
+    assert events[0].data["new_state"].attributes[ATTR_IN_PROGRESS] == 10
+    assert events[1].data["new_state"].attributes[ATTR_IN_PROGRESS] == 20
+    assert events[2].data["new_state"].attributes[ATTR_IN_PROGRESS] == 30
+    assert events[3].data["new_state"].attributes[ATTR_IN_PROGRESS] == 40
+    assert events[4].data["new_state"].attributes[ATTR_IN_PROGRESS] == 50
+    assert events[5].data["new_state"].attributes[ATTR_IN_PROGRESS] == 60
+    assert events[6].data["new_state"].attributes[ATTR_IN_PROGRESS] == 70
+    assert events[7].data["new_state"].attributes[ATTR_IN_PROGRESS] == 80
+    assert events[8].data["new_state"].attributes[ATTR_IN_PROGRESS] == 90
+    assert events[9].data["new_state"].attributes[ATTR_IN_PROGRESS] is False
+    assert events[9].data["new_state"].state == STATE_OFF
+
+
+async def test_update_with_progress_raising(hass: HomeAssistant) -> None:
+    """Test update with progress failing to install."""
+    state = hass.states.get("update.demo_update_with_progress")
+    assert state
+    assert state.state == STATE_ON
+    assert state.attributes[ATTR_IN_PROGRESS] is False
+
+    events = []
+    hass.helpers.event.async_track_state_change_event(
+        "update.demo_update_with_progress", callback(lambda event: events.append(event))
+    )
+
+    with patch(
+        "homeassistant.components.demo.update.asyncio.sleep",
```

Comment:
I'd introduce a "fake" object that represents the firmware library with a method call that normally awaits the sleep, which we can patch with a side effect.